hazelcast/serialization/compact.py

```diff
@@ -0,0 +1,2219 @@
+        default: typing.Optional[typing.List[typing.Optional[datetime.datetime]]],
+    ) -> typing.Optional[typing.List[typing.Optional[datetime.datetime]]]:
+        if not self._is_field_exists(field_name, FieldKind.ARRAY_OF_TIMESTAMP):
+            return default
+
+        return self.read_array_of_timestamp(field_name)
+
+    def read_array_of_timestamp_with_timezone(
+        self, field_name: str
+    ) -> typing.Optional[typing.List[typing.Optional[datetime.datetime]]]:
+        field = self._get_field_with_kind_check(
+            field_name, FieldKind.ARRAY_OF_TIMESTAMP_WITH_TIMEZONE
+        )
+        return self._read_var_sized_item_array(field, self._read_timestamp_with_timezone_helper)
+
+    def read_array_of_timestamp_with_timezone_or(
+        self,
+        field_name: str,
+        default: typing.Optional[typing.List[typing.Optional[datetime.datetime]]],
+    ) -> typing.Optional[typing.List[typing.Optional[datetime.datetime]]]:
+        if not self._is_field_exists(field_name, FieldKind.ARRAY_OF_TIMESTAMP_WITH_TIMEZONE):
+            return default
+
+        return self.read_array_of_timestamp_with_timezone(field_name)
+
+    def read_array_of_compact(
+        self, field_name: str
+    ) -> typing.Optional[typing.List[typing.Optional[typing.Any]]]:
+        field = self._get_field_with_kind_check(field_name, FieldKind.ARRAY_OF_COMPACT)
+        return self._read_var_sized_item_array(field, self._read_compact_helper)
+
+    def read_array_of_compact_or(
+        self, field_name: str, default: typing.Optional[typing.List[typing.Optional[typing.Any]]]
+    ) -> typing.Optional[typing.List[typing.Optional[typing.Any]]]:
+        if not self._is_field_exists(field_name, FieldKind.ARRAY_OF_COMPACT):
+            return default
+
+        return self.read_array_of_compact(field_name)
+
+    def _get_field(self, field_name: str) -> "FieldDescriptor":
+        field = self._schema.fields.get(field_name)
+        if not field:
+            raise HazelcastSerializationError(
+                f"No field with the name '{field_name}' can found in the {self._schema}"
+            )
+
+        return field
+
+    def _get_field_with_kind_check(
+        self, field_name: str, field_kind: "FieldKind"
+    ) -> "FieldDescriptor":
+        field = self._get_field(field_name)
+        if field.kind != field_kind:
+            DefaultCompactReader._raise_mismatched_field_kind_error(
+                field_name, field.kind, field_kind
+            )
+
+        return field
+
+    def _read_boolean_internal(self, field: "FieldDescriptor") -> bool:
+        position = self._data_start_position + field.position
+        bit_position = field.bit_position
+        packed_byte = self._inp.read_byte_positional(position)
+        return ((packed_byte >> bit_position) & 0x01) != 0
+
+    @staticmethod
+    def _raise_mismatched_field_kind_error(
+        field_name: str, field_kind: "FieldKind", expected_field_kind: "FieldKind"
+    ) -> None:
```

Comment:
What about using `typing.NoReturn` for methods that always raise? BTW, How are you using the type hints? Are they useful in IDE or mypy? I am seeing a lot of warnings in mypy but not sure if they are important.

Reply:
After merging 535, we will have non-strict mypy checks in the PR builder